src/sat/sat_clause.cpp

```diff
@@ -123,53 +123,80 @@ namespace sat {
 
     clause_allocator::clause_allocator():
         m_allocator("clause-allocator") {
-#ifdef _AMD64_
+#if defined(_AMD64_) 
         m_num_segments = 0;
 #endif
     }
 
     clause * clause_allocator::get_clause(clause_offset cls_off) const {
-#ifdef _AMD64_
-        return reinterpret_cast<clause *>(m_segments[cls_off & c_aligment_mask] + (static_cast<size_t>(cls_off) & ~c_aligment_mask));
+#if defined(_AMD64_) 
+#if defined (Z3DEBUG)
+        clause const* result;
+        if (((cls_off & c_alignment_mask) == c_last_segment)) { 
+            unsigned id = cls_off >> c_cls_alignment;
+            bool check = m_last_seg_id2cls.find(id, result);
+            SASSERT(check);
+            return const_cast<clause*>(result);
+        }
+#endif
+        return reinterpret_cast<clause *>(m_segments[cls_off & c_alignment_mask] + (static_cast<size_t>(cls_off) & ~c_alignment_mask));
```

Comment:
Again could really do with a comment here explaining we are mapping a 32-bit integer to a `clause*`.